File: MoonEngine/Render/src/main/java/com/greentree/engine/moon/render/assets/shader/ShaderDataAssetSerializator.java
package com.greentree.engine.moon.render.assets.shader;

import com.greentree.engine.moon.assets.Value1Function;
import com.greentree.engine.moon.assets.asset.Asset;
import com.greentree.engine.moon.assets.asset.AssetKt;
import com.greentree.engine.moon.assets.key.AssetKey;
import com.greentree.engine.moon.assets.serializator.AssetSerializator;
import com.greentree.engine.moon.assets.serializator.loader.AssetLoader;
import com.greentree.engine.moon.render.shader.ShaderDataImpl;
import com.greentree.engine.moon.render.shader.ShaderLanguage;
import com.greentree.engine.moon.render.shader.ShaderType;

public class ShaderDataAssetSerializator implements AssetSerializator<ShaderDataImpl> {

    @Override
    public Asset<ShaderDataImpl> load(AssetLoader.Context manager, AssetKey ckey) {
        if (ckey instanceof ShaderAssetKey key) {
            final var text = manager.load(String.class, key.text());
            final var type = key.shaderType();
            final var language = key.language();
            return AssetKt.map(text, new TextToShaderDataFunction(type, language));
        }
        return null;
    }

    private record TextToShaderDataFunction(ShaderType type, ShaderLanguage language)
            implements Value1Function<String, ShaderDataImpl> {

        private static final long serialVersionUID = 1L;

        @Override
        public String toString() {
            return "TextToShaderData[type=" + type + ", language=" + language + ']';
        }

        @Override
        public ShaderDataImpl apply(String value) {
            if (value == null)
                return null;
            return new ShaderDataImpl(value, type, language);
        }

    }

}

File: MoonEngine/Render/src/main/java/com/greentree/engine/moon/render/assets/shader/ShaderDataAssetSerializator.kt
package com.greentree.engine.moon.render.assets.shader

import com.greentree.engine.moon.assets.NotSupportedKeyType
import com.greentree.engine.moon.assets.key.AssetKey
import com.greentree.engine.moon.assets.loader.AssetLoader
import com.greentree.engine.moon.assets.loader.load
import com.greentree.engine.moon.assets.serializator.AssetSerializator
import com.greentree.engine.moon.render.shader.ShaderDataImpl
import com.greentree.engine.moon.render.shader.ShaderLanguage
import com.greentree.engine.moon.render.shader.ShaderType

data object ShaderDataAssetSerializator : AssetSerializator<ShaderDataImpl> {

	override fun load(manager: AssetLoader.Context, ckey: AssetKey): ShaderDataImpl {
		if(ckey is ShaderAssetKey) {
			val text = manager.load<String>(ckey.text)
			val type: ShaderType = ckey.shaderType
			val language: ShaderLanguage = ckey.language
			return ShaderDataImpl(text, type, language)
		}
		throw NotSupportedKeyType
	}
}
